packages/composites/proof-of-reserves/src/utils/addressValidator.ts

```diff
@@ -0,0 +1,209 @@
+  }
+  return validatedAddresses
+}
+
+/**
+ * Returns either a valid Ethereum-style address with a valid checksum
+ * or logs a warning and returns undefined
+ */
+const getValidEvmAddress = (address: string): string | undefined => {
+  try {
+    return utils.getAddress(address)
+  } catch (error) {
+    Logger.warn(
+      error,
+      `The address "${address}" is invalid or has an invalid checksum and has been removed from the request.`,
+    )
+  }
+  return
+}
+
+const getValidBtcAddress = (address: string): string | undefined => {
+  const addressPrefix = address[0]
+  switch (addressPrefix) {
+    // Legacy (P2PKH) and Nested SegWit (P2SH) Bitcoin addresses start with 1 and are case-sensitive
+    case '1':
+    case '3':
+      if (address.length === 34 && isBase58(address)) return address
+      Logger.warn(
+        { warning: 'Invalid address detected' },
+        `The address "${address}" is not a valid Bitcoin address and has been removed.`,
+      )
+      return
+    case 'b':
+    case 'B':
+      address = address.toLowerCase()
+      if (address.slice(0, 3) === 'bc1' && address.length === 42 && isBech32(address.slice(3)))
+        return address
+      Logger.warn(
+        { warning: 'Invalid address detected' },
+        `The address "${address}" is not a valid Bitcoin address and has been removed.`,
+      )
+      return
+    default:
+      Logger.warn(
+        { warning: 'Invalid address detected' },
+        `The address "${address}" is not a valid Bitcoin address and has been removed.`,
+      )
+      return
+  }
+}
+
+const getValidDogeAddress = (address: string): string | undefined => {
+  if (address[0] !== 'D') return
+  if (isBase58(address.slice(1))) return address
+  Logger.warn(
+    { warning: 'Invalid address detected' },
+    `The address "${address}" is not a valid Dogecoin address and has been removed.`,
+  )
+  return
+}
+
+const getValidCardanoAddress = (address: string): string | undefined => {
+  if (address.slice(0, 4).toLowerCase() === 'addr' && isBech32(address.slice(5).toLowerCase()))
+    return address.toLowerCase()
+  if (isBase58(address)) return address
+  return
+}
+
+const getValidFilecoinAddress = (address: string): string | undefined => {
+  address = address.toLowerCase()
+  if (address[0] !== 'f' && address[0] !== 't') return
+  if (address[1] !== '1' && address[1] !== '3') return
+  if (isBase32(address.slice(2))) return address
+  return
+}
+
+export const filterDuplicates = (addresses: AddressObject[]): AddressObject[] => {
+  const uniqueMap: Record<string, boolean> = {}
+  const uniqueAddresses: AddressObject[] = []
+  for (const addressObject of addresses) {
+    if (uniqueMap[addressObject.address]) {
+      Logger.warn(
+        { warning: 'Duplicate address detected' },
+        `The address "${addressObject.address}" is duplicated in the request and the duplicate has been removed.`,
```

Comment:
May want to also include the job run ID to help link this warning to the job that is throwing it.

Reply:
Good call on that. I will add it